### ISSUE TITLE
Mismatch between `__init__` arg name and matching attribute results in InternalError in 3.10 pattern matching

### ISSUE THREAD

**Issue by marcinplatek** [microsoft/pylance-release#3197]
## Environment data
-   Language Server version: v2022.8.20
-   OS and version: macOS Monterey 12.4
-   Python version: 3.10.5

## Code Snippet

```python
class Token:
    def __init__(self, type_: str) -> None:
        self.type = type_


t = Token(type_="TEST")

match t:
    case Token(type="TEST"):
        pass
    case _:
        pass
```

## Repro Steps

EDIT: set pylance to strict

Try linting the code snippet. 

The original error occurred in Lark library (https://github.com/lark-parser/lark/issues/1180).

## Expected behavior

Either a regular typing error or no error at all.

## Actual behavior

An internal error that breaks the typing completely, even if I add # type: ignore

## Logs

```
An internal error occurred while type checking file ".../bug_example.py": TypeError: Cannot read properties of undefined (reading 'length')
    at Object.isPrivateName (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/symbolNameUtils.ts:16:17)
    at O._conditionallyReportPrivateUsage (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/checker.ts:3435:47)
    at O.visitName (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/checker.ts:1334:14)
    at O.visitNode (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:261:29)
    at O.walk (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:98:37)
    at O.walk (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/checker.ts:280:19)
    at forEach (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:107:22)
    at Array.forEach (<anonymous>)
    at O.walkMultiple (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:105:15)
    at O.walk (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:100:18)
    at O.walk (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/checker.ts:280:19)
    at forEach (.../.vscode/extensions/ms-python.vscode-pylance-2022.8.20/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:107:22)
    at Array.forEach (<anonymous>)```

**Comment by marcinplatek** [comment]
Oh, my bad, I've forgotten about setting pylance to strict.

The error should look like that:

<img width="257" alt="Screenshot 2022-08-15 at 18 10 18" src="https://user-images.githubusercontent.com/23441199/184672854-fe879848-0783-4e50-8a21-3f89f6393d39.png">

<img width="440" alt="Screenshot 2022-08-15 at 18 11 18" src="https://user-images.githubusercontent.com/23441199/184672805-ab3da6df-786b-4670-a613-ce0832192b00.png">

**Comment by erictraut** [comment]
Thanks, I'm able to repro. This will be fixed in the next release.

**Comment by erictraut** [comment]
Here are some additional details about this bug for those who are interested.

Python 3.10 introduced the first so-called "soft keywords" to the Python grammar: `match` and `case`. These are treated as keywords when they're used in certain contexts but as identifiers in all other situations. Contrast that with "hard keywords" like `def` and `class` which are illegal identifiers.

In [draft PEP 695](https://peps.python.org/pep-0695/), I proposed to add another soft keyword to Python 3.12: `type`. The latest version of pyright (and therefore pylance, which builds on pyright) includes a prototype of the draft PEP.

This bug was caused by some incorrect logic in pyright's parser when dealing with soft keywords that appear within named arguments for class patterns. You will see the same problem if you use `match`, `case`, or `type`.
